Update geopmaccess man page and CLI

cmcantalupo opened the pull request:
- Fixes #2212
- Relates to #2198
- Could have better negative testing for incompatible command line option errors.
- Does not implement the -d option described in #2198 to remove a subset of names from an access list, added --edit mode instead
- Solution for #2212 is not exactly as described in the issue, but hopefully is sufficient.

cmcantalupo committed:
Get rid of m2r kruft

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add a man page for geopmaccess(1)

- Defines interface to support #2212

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Update geopmaccess CLI to match changes to man page

- Add -u/--user command line option
- Add -D/--delete command line option
- Add -n/--dry-run command line option
- add -E/--skip-check command line option

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented:
@bensallen 

This is currently a draft PR since the tests are broken.  It also does not provide exactly the solutions described in the associated issues.  At this time, feedback on the man page here:

https://github.com/cmcantalupo/geopm/blob/geopmaccess-cli/service/docs/source/geopmaccess.1.rst

is appreciated.  I'm hoping to get this out of draft status tomorrow or Wednesday.  Thanks!

cmcantalupo committed:
Update geopmaccess man page based on implementation requirements

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix tests and issue shown by test

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

dannosliwcd reviewed (commented):
I noticed `geopmaccess` is installed even when using `--disable-systemd`. Should it be skipped in that case?

dannosliwcd commented on service/docs/source/geopmaccess.1.rst:
```diff
@@ -1,28 +1,181 @@
+  files
+
+*
+  ``-E``, ``--skip-check``:
```
Maybe `-q` `--quiet`?

dannosliwcd replied:
Nevermind. Reached the description at the end of the man page, and I see this is different from a quiet mode. Maybe `-f`, `--force` fits other CLIs that allow similar validation skips.

dannosliwcd commented on service/docs/source/geopmaccess.1.rst:
```diff
@@ -1,28 +1,181 @@
+  Command applies to controls not signals.
+
+
+* ``-u``, ``--user``:
```
`-u` is missing from the synopsis. Given that there's a `-g GID`/`--group GID` it seems unconventional to have a `-u`/`--user` that doesn't accept UID. Maybe `--default` instead?

dannosliwcd commented on service/docs/source/geopmaccess.1.rst:
```diff
@@ -1,28 +1,181 @@
+The administrator may execute ``geopmaccess`` to write to an access
+list by providing the ``-w`` / ``--write`` command line option.  The
+``-D`` / ``--delete`` option will remove all signals or controls from
+the configuration.
```
The default behavior is different in reads vs writes. In reads, it defaults to the signals of the calling process. In writes/deletes it defaults to the user-default list. Maybe make the difference stand out by ending this paragraph with the reading default behavior, to mirror the end of the writing paragraph.

dannosliwcd commented:
Regarding "remove a subset of names from an access list", one option could be to open the list-to-be-modified in an editor (e.g., `-e` `--editor <given editor or default to $EDITOR>`. If the file can't be modified in place (e.g., since it needs to be validated), we could prime a temporary buffer with the current state, then do our work on the written state (like the default behavior on git commit messages).

Krebrov001 commented on service/docs/source/geopmaccess.1.rst:
```diff
@@ -1,28 +1,181 @@
+
+The access list created is derived from standard input by reading one
+name per line.  Standard input lines that begin with the ``#``
+character are ignored, and an empty line or EOF (end of file) will
```
```rst
character are ignored, and an empty line or ``EOF`` *(end of file)* will
```

Krebrov001 replied:
Please add this formatting.

Krebrov001 reviewed (dismissed):
Looks good, I like it.

dannosliwcd commented on service/docs/source/geopmaccess.1.rst:
```diff
@@ -1,28 +1,181 @@
 -----------
 
-Access management for the geopm service.
+The GEOPM service uses the ``/etc/geopm-service`` directory to store
+files that control which signals and controls may be accessed by a
+user through the service.  The purpose of the ``geopmaccess`` command
+line tool is to enable reading and writing of these access list files.
```
One of the described use cases is for config validation. Maybe add a section or paragraph on failure scenarios.

I don't think it is necessary to enumerate every failure case (though I'm not against that if you want it). It should be sufficient to say what the user should expect to happen in pass/fail cases though. E.g., does it say that an error occurred, or what input caused the error? Does it set the process return code based on pass/fail state? How do options (e.g., dry-run or skip-check) impact this behavior?

cmcantalupo committed:
Address review comments

- Add a --edit mode that allows the use of text editor to modify access lists
- Change --user to --default and --skip-check to --force
- TODO: Need to add a section to the man page on errors and return value.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented:
@dannosliwcd The --disable-systemd configure option seems to be confusing for most people who use it.  Perhaps it should be something like --disable-service-io-group since this is really the only thing this option does.  We require libsystemd.so specifically to call into systemd DBus from C++ to support the ServiceIOGroup, and for no other reason.  

I'll complete the EXIT STATUS section of the man page soon.  Thanks for all the helpful suggestions.

cmcantalupo commented:
Still lacking unit tests for all these new features added to the command line interface, but a preliminary by hand integration testing shows geopmaccess functioning as expected (including edit mode).

cmcantalupo committed:
Refactor run implementation for clarity

- Put all error checking at the top of the method.
- Factor out editing int separate function
- Reduce branching

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add an example to the geopmaccess man page

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Move examples in GEOPM Service admin guide

- Move these geopmaccess examples into the
  geopmaccess man page.
- The admin guide needs more work.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Use long form options in man page examples and add exit status section

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Some clean up of geopmaccess man page

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add unit tests for new command line options for geopmaccess

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Fix default behavior for geopmaccess to match man page

- The default behavior is now to print the calling
  user's access.
- The default access list is accessed with the -u/--default
  option now

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo commented:
Unit tests have been added for the new features.  This included fixing the implementation to match man page.  Some negative testing could still be added to hit the exceptions at the top of the run command.

dannosliwcd commented on service/docs/source/geopmaccess.1.rst:
```diff
@@ -170,6 +190,215 @@ Note that having signal or control names in an access list in
+will occur otherwise.
+
+Some command line options cannot be used together, for example, a
+failures will occur if the user specifies both ``--all`` and
```
```suggestion
failure will occur if the user specifies both ``--all`` and
```

dannosliwcd commented on service/geopmdpy/access.py:
```diff
@@ -207,13 +250,46 @@ def get_group_controls(self, group):
+            initial_names = self.get_group_controls(group)
+        else:
+            initial_names = self.get_group_signals(group)
+        edit_path = self.edited_names_path(initial_names)
+        try:
+            with open(edit_path) as fid:
+                names = self.read_names(fid)
+        finally:
+            os.unlink(edit_path)
```
Could potentially avoid manual `os.unlink` and `os.close` calls here and in `edited_names_path` if this block is instead wrapped in a `with tempfile.NamedTemporaryFile()` context and the tempfile.name path is passed as an argument to `edited_names_path`.

cmcantalupo replied:
I had some trouble implementing that without having the file open in python while the editor also had it open.  Because the unlink() is bound to the close() call when you use the ``__enter__()``/``__exit__()`` interfaces through ``with`` you always get both calls.  In this case we would like to close() the file in python without unlink() while the editor process modifies the file.  After the file is open() again in python we will unlink() it when we close() the second time.

cmcantalupo replied:
Okay, no need to close the file while it is edited.  Thanks for the suggestion.

cmcantalupo committed:
Apply suggested change to documentation from Daniel

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
Co-authored-by: Daniel Wilson <daniel1.wilson@intel.com>

cmcantalupo committed:
Fix dead link in runtime.html

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Simplfy the way temporary file is handled when edit mode is used

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Clean up first refefence to capabilities

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>

cmcantalupo committed:
Add some more content to the admin page

- Could use some work, but at least there is
  some content now that the examples have all
  been moved.

Signed-off-by: Christopher M. Cantalupo <christopher.m.cantalupo@intel.com>
